streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -235,8 +350,67 @@ private[streaming] class ReceivedBlockTracker(
     }
   }
 
+  /**
+   * Creates a WAL Writer in a separate thread to enable batching of log events.
+   * Exposed for tests.
+   */
+  protected def createBatchWriteAheadLogWriter(): Option[BatchLogWriter] = {
+    if (!WriteAheadLogUtils.isBatchingEnabled(conf)) return None
+    val writer = checkpointDirOption.map(_ => new BatchLogWriter)
+    writer.foreach { runnable =>
+      new Thread(runnable, "Batch WAL Writer").start()
+      Runtime.getRuntime.addShutdownHook(new Thread {
+        override def run(): Unit = runnable.stop()
+      })
+    }
+    writer
+  }
+
   /** Check if the write ahead log is enabled. This is only used for testing purposes. */
   private[streaming] def isWriteAheadLogEnabled: Boolean = writeAheadLogOption.nonEmpty
+
+  /** A helper class that writes LogEvents in a separate thread to allow for batching. */
+  private[streaming] class BatchLogWriter extends Runnable {
+
+    var active: Boolean = true
+
+    private def writeRecords(records: List[ReceivedBlockTrackerLogEvent]): Unit = {
+      writeAheadLogOption.foreach { logManager =>
+        if (records.nonEmpty) {
+          logDebug(s"Batched ${records.length} records for WAL write")
+          logManager.write(ByteBuffer.wrap(Utils.serialize(CombinedRBTLogEvent(records))), 
+            clock.getTimeMillis())
+        }
+      }
+    }
+
+    def stop(): Unit = {
+      active = false
+    }
+
+    private def flushRecords(): Unit = {
+      val buffer = new ArrayBuffer[ReceivedBlockTrackerLogEvent]()
+      while (!walWriteQueue.isEmpty) {
+        buffer.append(walWriteQueue.poll())
+      }
+      val records = buffer.toList
+      try {
+        writeRecords(records)
+        records.foreach(walWriteStatusMap.put(_, Success))
+      } catch {
+        case e: Exception =>
+          logWarning(s"Batch WAL Writer failed to write $records")
+          records.foreach(walWriteStatusMap.put(_, Fail))
+      }
+    }
+
+    override def run(): Unit = {
+      while (active) {
+        flushRecords()
+        Thread.sleep(WAL_WRITE_STATUS_CHECK_BACKOFF)
```

Comment:
I am a little afraid that if there is an exception in this loop, the thread with exit and nothing will get synced. I think it may be a good idea to have another try catch around the whole thing in run().